linked_list/linked_list.py

```diff
@@ -13,64 +14,170 @@ def __init__(self):
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
     def add_first(self, value):
-        pass
+
+        # head POINTS to a value the way next does
+        # if head is not pointing anywhere, make head point to new_node
+
+        new_node = Node(value, next_node=self.head)
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        
+        current = self.head
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        list_length = 0
+
+        current = self.head
+        while current != None:
+            list_length += 1
+            current = current.next
+        
+        return list_length
+        
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if self.head == None:
+            return None
+
+        # if the index is 0, return value head is pointing to
+        if index == 0:
+            return self.head.value
+        
+        current = self.head
+        current_index = 0
+        while current.next and current_index < index - 1: # stop at node just before the one specified
+            current = current.next
+            current_index += 1
+
+        # read the value in the next node
+        if current.next:
+            return current.next.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        while current != None:
+            if current.next == None:
+                return current.value
+            current = current.next
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1) based on insertion operation, but O(n) based on traversing?
+    # Space Complexity: O(n)
     def add_last(self, value):

```

Comment:
👍 However the time complexity is O(n) and space O(1) since you're only creating ONE node